[material-ui] Add `DefaultPropsProvider` for Pigment CSS integration

siriwatknp opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

closes #42635 

The gist is https://github.com/mui/material-ui/pull/42638/files#diff-532de7437125015b5ab87338e5e5950bc513e99fc7d987838a0f3755ee01408b, the rests are replacing `createUseThemeProps` with `useDefaultProps`.

The signature changes but still support v5:

```js
<DefaultPropsProvider
  value={{
    MuiButton: {
      disableTouchRipple: true,
    }
  }}
>
```

## Testing

Only the spec is added because the logic is already tested by `describeConformance` from all the components.

---

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/material-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

mnajdova reviewed (commented):
We have a bit of misalignment in the names:

- `PropsProvider` vs `useDefaultProps` - we should either have `DefaultPropsProvider` and `useDefaultProps`, or `PropsProvider` and `useProps`


I like that the API is identical as before, but in ideal scenario this API in my opinion should be:

```
<DefaultPropsProvider
  value={{
    MuiButton: { disableRipple: true }, // note that I removed the defaultProps key
  }}
>
```

I would prefer this API ☝️

siriwatknp commented:
> We have a bit of misalignment in the names:
> 
> * `PropsProvider` vs `useDefaultProps` - we should either have `DefaultPropsProvider` and `useDefaultProps`, or `PropsProvider` and `useProps`
> 
> I like that the API is identical as before, but in ideal scenario this API in my opinion should be:
> 
> ```
> <DefaultPropsProvider
>   value={{
>     MuiButton: { disableRipple: true }, // note that I removed the defaultProps key
>   }}
> >
> ```
> 
> I would prefer this API ☝️

Changed to `DefaultPropsProvider` and support both signatures.

siriwatknp commented on packages/mui-codemod/src/deprecations/tab-classes/test-cases/expected.css:
```diff
@@ -1,3 +1,4 @@
 .MuiTab-root .MuiTab-icon {
   color: red;
 }
+
```
This is not related to this PR. I have no idea how this passed the check before merging to `next`.

DiegoAndai replied:
Fixed in https://github.com/mui/material-ui/pull/42686

mnajdova commented on packages/mui-system/src/DefaultPropsProvider/DefaultPropsProvider.tsx:
```diff
@@ -0,0 +1,61 @@
+  }
+
+  if (!config.styleOverrides && !config.variants) {
+    // v6 signature, no property 'defaultProps'
```
Good call for adding both APIs 👌

mnajdova reviewed (approved):
Nice work, let's merge this before the beta 🎉 This was the last missing part for interoperability, right?

siriwatknp commented:
> Nice work, let's merge this before the beta 🎉 This was the last missing part for interoperability, right?

No. There are 2 more, [left over components](https://github.com/mui/material-ui/pull/42640) and Pigment layout components (need decision on https://github.com/mui/pigment-css/pull/144#issuecomment-2179809493)

oliviertassinari commented on packages/mui-system/src/DefaultPropsProvider/DefaultPropsProvider.tsx:
```diff
@@ -0,0 +1,61 @@
+import PropTypes from 'prop-types';
+import resolveProps from '@mui/utils/resolveProps';
+
+const PropsContext = React.createContext<Record<string, any> | undefined>(undefined);
```
We are missing the codebase guideline for React context here:

```suggestion
const PropsContext = React.createContext<Record<string, any> | undefined>(undefined);

if (process.env.NODE_ENV !== 'production') {
  PropsContext.displayName = 'DefaultProps';
```

oliviertassinari replied:
Also, the name of the context looks wrong

```suggestion
const DefaultPropsContext = React.createContext<Record<string, any> | undefined>(undefined);
```
